atoma-state/src/state_manager.rs

```diff
@@ -2902,6 +2910,66 @@ impl AtomaState {
         Ok(())
     }
 
+    /// Records or updates a key rotation event for a specific epoch in the database.
+    ///
+    /// This method tracks cryptographic key rotation events by maintaining a counter for each epoch.
+    /// It uses an "upsert" operation, meaning it will either:
+    /// - Insert a new record if no rotation exists for the given epoch
+    /// - Update the existing record if a rotation was already recorded for that epoch
+    ///
+    /// # Arguments
+    ///
+    /// * `epoch` - The epoch number when the key rotation occurred
+    /// * `key_rotation_counter` - The cumulative number of key rotations that have occurred
+    ///
+    /// # Returns
+    ///
+    /// - `Result<()>`: A result indicating success (Ok(())) or failure (Err(AtomaStateManagerError))
+    ///
+    /// # Errors
+    ///
+    /// This function will return an error if:
+    /// - The database query fails to execute
+    /// - There's a connection issue with the database
+    ///
+    /// # Example
+    ///
+    /// ```rust,ignore
+    /// use atoma_node::atoma_state::AtomaStateManager;
+    ///
+    /// async fn rotate_keys(state_manager: &AtomaStateManager) -> Result<(), AtomaStateManagerError> {
+    ///     let current_epoch = 100;
+    ///     let rotation_counter = 5;
+    ///     
+    ///     state_manager.insert_new_key_rotation(current_epoch, rotation_counter).await
+    /// }
+    /// ```
+    ///
+    /// # Security Considerations
+    ///
+    /// This method is part of the system's cryptographic key management infrastructure and helps:
+    /// - Maintain an audit trail of key rotation events
+    /// - Verify compliance with key rotation policies
+    /// - Track the frequency of key rotations per epoch
+    #[instrument(level = "trace", skip_all, fields(%epoch, %key_rotation_counter))]
+    pub async fn insert_new_key_rotation(
+        &self,
+        epoch: i64,
+        key_rotation_counter: i64,
+    ) -> Result<()> {
+        sqlx::query(
+            "INSERT INTO key_rotations (epoch, key_rotation_counter) VALUES ($1, $2)
+            ON CONFLICT (epoch)
+            DO UPDATE SET epoch = EXCLUDED.epoch,
```

Comment:
you don't need to set the epoch on conflict on epoch, it's equal already

Reply:
Addressed !